src/libstd/sys/unix/stack_overflow.rs

```diff
@@ -137,11 +137,16 @@ mod imp {
     }
 
     unsafe fn get_stackp() -> *mut libc::c_void {
+        let page_size = crate::sys::unix::os::page_size();
         let stackp =
-            mmap(ptr::null_mut(), SIGSTKSZ, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+            mmap(ptr::null_mut(), SIGSTKSZ + page_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
         if stackp == MAP_FAILED {
             panic!("failed to allocate an alternative stack");
         }
+        let guard_result = libc::mprotect(stackp, page_size, PROT_NONE);
+        if guard_result != 0 {
+            panic!("failed to set up alternative stack guard page");
+        }
         stackp
```

Comment:
You're returning the protected page as the bottom of the stack. That will sort of work as long as `SIGSTKSZ` is larger than the page size to begin with, so the top will still be in R/W memory, but I think you want to return `stackp.add(page_size)`.

The `drop_handler` also needs to be adjusted to account for the correct map size.

Reply:
I'd entirely neglected `drop_handler`! Thank you. I've adjusted `get_stackp` to return `stackp.add(page_size)` as well.